### DISCUSSION TITLE
What is the way to perform async pagination using 2.0

### DISCUSSION THREAD

**Discussion by Junkkk** [sqlalchemy/sqlalchemy#10254] [Usage Questions]
I'm trying to write a cool function to get result which can be paginated

this is my weak example

```python
import asyncio
import uuid

from sqlalchemy import Column, ForeignKey, select, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker, AsyncSession
from sqlalchemy.orm import Mapped, relationship, declarative_base

Base = declarative_base()


class User(Base):
    __tablename__ = 'users'

    uuid = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str]
    type_uuid = Column(UUID(as_uuid=True), ForeignKey("user_types.uuid"), nullable=False)

    user_type: Mapped["UserType"] = relationship(backref="users")


class UserType(Base):
    __tablename__ = 'user_types'

    uuid = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str]


db_url = "postgresql+asyncpg://example_app:examplepass@localhost:5433/example_app_db_test"
engine = create_async_engine(db_url)
AsyncSessionFactory = async_sessionmaker(
    engine, expire_on_commit=False
)


async def get_paginated(session: AsyncSession, limit: int, offset: int):
    q = (
        select(User)
        .join(UserType)
        .filter(UserType.name == 'some_type1')
    )
    total_q = (
        select(func.count(User.uuid))
        .select_from(q)
    )
    total_count = await session.scalar(total_q)
    users = await session.scalars(q.limit(limit).offset(offset))

    return total_count, users


async def async_main():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSessionFactory() as session:
        user_type1 = UserType(name='some_type1')
        user_type2 = UserType(name='some_type2')
        user_type3 = UserType(name='some_type3')
        session.add_all([user_type1, user_type2, user_type3])
        await session.commit()

        user1 = User(name='user1', type_uuid=user_type1.uuid)
        user2 = User(name='user2', type_uuid=user_type2.uuid)
        user3 = User(name='user3', type_uuid=user_type1.uuid)
        user4 = User(name='user4', type_uuid=user_type2.uuid)
        user5 = User(name='user5', type_uuid=user_type3.uuid)
        user6 = User(name='user6', type_uuid=user_type1.uuid)
        session.add_all([user1, user2, user3, user4, user5, user6])
        await session.commit()

        total, users = await get_paginated(session, limit=10, offset=0)
        print(f"{total = }")
        for user in users:
            print(f"{user.name = }")

asyncio.run(async_main())
```

after executing this i got deprecation warning and incorrect result of total(users correct)

```
SADeprecationWarning: Implicit coercion of SELECT and textual SELECT constructs into FROM clauses is deprecated; please call .subquery() on any Core select or ORM Query object in order to produce a subquery object.
  .select_from(q)
sys:1: SAWarning: SELECT statement has a cartesian product between FROM element(s) "anon_1" and FROM element "users".  Apply join condition(s) between each element to resolve.
total = 18
user.name = 'user1'
user.name = 'user3'
user.name = 'user6'
```

actually I want to ask the best way to solve it 

Thanks for attention. don't judge, my first discussion

**Comment by zzzeek** [comment]
the warning is because you are feeding a `select()` into `select_from()` as though it's a subquery, which it isn't technically until you call `.subquery()` on it

```py
async def get_paginated(session: AsyncSession, limit: int, offset: int):
    q = (
        select(User)
        .join(UserType)
        .filter(UserType.name == 'some_type1')
    )
    total_q = (
        select(func.count(User.uuid))
        .select_from(q.subquery())
    )
```

the next thing that you'd want to address is the use of `offset()`, which if you have more than a few thousand rows will very quickly slow things way down, as the DB needs to scan the whole rowset up til that offset.     So there's a pretty well known blog / presentation /whatever about the OFFSET thing worth going through which is at https://use-the-index-luke.com/blog/2013-07/pagination-done-the-postgresql-way  , where he gets at the main thing we want to do is to select from regions of the table directly based on some other criteria, like dates or whatever it is you are ordering by.  There is also https://use-the-index-luke.com/sql/partial-results/fetch-next-page that talks about this.

We have some variations on this theme in the wiki which make use of SQL window functions, which can be pretty useful for giving you segments of a table while at the same time giving you some kind of countable index, I'd check that out at https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery.    again the use-the-index blog has some ideas for this at https://use-the-index-luke.com/sql/partial-results/window-functions .

**Reply by CaselIT** [reply]
Any way you choose you will need to add an order to your query though, otherwise you are returning meaningless data since a select without order can return data in order way the db likes (including a different one for the same query run twice)

**Reply by Junkkk** [reply]
thank you for a very quick response, I will try to comprehend this material. But I don't understand why I have total_count = 18 and 3 users in result. This queries looks same

**Reply by CaselIT** [reply]
> SELECT statement has a cartesian product between FROM element(s) "anon_1" and FROM element "users".  Apply join condition(s) between each element to resolve.

you have this warning, so it seems that the total query gives you a cartesian product. It's probably because you have `func.count(User.uuid)` but you select from a subquery, so there will be `from (the subquery), user`
You can most simply use `func.count('*')` or `func.count(1)`
